### ISSUE TITLE
Prevent UnitTasks from continuing after exiting play mode

### ISSUE THREAD

**Issue by joelmalone** [Cysharp/UniTask#22]
`UniTask`s continue executing after leaving play mode.  This may lead to issues where continuations are executed after returning to edit mode, potentially altering the scene.

Note that this is the same behaviour as the built-in `Task`s, which makes them extremely risky and essentially only reliable in edit mode, thus only useful for editor tooling.

It'd be nice if `UniTask`s wouldn't run after play mode exits.

Here's a simple example case that represents the problem.  Add this MonoBehaviour to a game object, hit the Play button, then immediately exit play mode.  You'll notice _some_ of the instantiated game objects are destroyed when returning to edit mode, but the UniTask keeps running and instantiates more _into the edit mode scene_, thus permanently changing the scene.

```cs
using System;
using UniRx.Async;
using UnityEngine;

public class UniTaskTest : MonoBehaviour
{

    protected async UniTask Start()
    {
        await InstantiateObjectsForTwoSeconds();
    }

    async UniTask InstantiateObjectsForTwoSeconds()
    {
        var end = DateTimeOffset.Now.AddSeconds(2);

        var counter = 0;
        while (DateTimeOffset.Now < end)
        {
            await UniTask.Yield();

            var go = GameObject.CreatePrimitive(PrimitiveType.Cube);
            go.transform.position = new Vector3(0, 0, counter++);
        }
    }

}
```

A _very naive_ hack is to modify `PlayerLoopHelper.InsertRunner()` so it only calls `ContinuationQueue.Run()` and `PlayerLoopRunner.Run()` when in play mode, e.g.:

```cs
        static PlayerLoopSystem[] InsertRunner(PlayerLoopSystem loopSystem, Type loopRunnerYieldType, ContinuationQueue cq, Type loopRunnerType, PlayerLoopRunner runner)
        {
            var yieldLoop = new PlayerLoopSystem
            {
                type = loopRunnerYieldType,
                updateDelegate = () =>
                {
                    // This line is the new edit-mode check
                    if (Application.isPlaying)
                        cq.Run();
                }
            };

            var runnerLoop = new PlayerLoopSystem
            {
                type = loopRunnerType,
                updateDelegate = () =>
                {
                    // This line is the new edit-mode check
                    if (Application.isPlaying)
                        runner.Run();
                }
            };

            // (the remaining code is unmodified)

            var dest = new PlayerLoopSystem[loopSystem.subSystemList.Length + 2];
            Array.Copy(loopSystem.subSystemList, 0, dest, 2, loopSystem.subSystemList.Length);
            dest[0] = yieldLoop;
            dest[1] = runnerLoop;
            return dest;
        }
```

I would love to hear some thoughts on this, and would be willing to spend some time solving the general issue of play mode Tasks executing in edit mode - right now the default behaviour of `Task`s and `UniTask`s makes them generally unsuitable for use in Unity projects.

**Comment by neuecc** [comment]
Thanks, it seems good.
If finalization is defined, there is a concern that it will never happen.

**Comment by neuecc** [comment]
I've implemented as your suggested, thank you!
and released `Ver 1.1.0`.

**Comment by Thaina** [comment]
@neuecc Sadly now I want `UniTask` to run in edit mode too. I have a dynamic `GameObject` generator that could download texture and generate sprite in both play mode and prefab mode. And if I need to rely on `UniTask` to sync with `PlayerLoop` it will stop working in editor

Maybe it should be like this instead?

```C#

		static PlayerLoopSystem[] InsertRunner(PlayerLoopSystem loopSystem,Type loopRunnerYieldType,ContinuationQueue cq,Type loopRunnerType,PlayerLoopRunner runner)
		{
// please implement Clear() function for PlayerLoopRunner and ContinuationQueue
			UnityEditor.EditorApplication.playModeStateChanged += (state) => {
				if(runner != null)
					runner.Clear();
				if(cq != null)
					cq.Clear();
			};

			var yieldLoop = new PlayerLoopSystem
			{
				type = loopRunnerYieldType,
				updateDelegate = cq.Run
			};

			var runnerLoop = new PlayerLoopSystem
			{
				type = loopRunnerType,
				updateDelegate = runner.Run
			};

			var dest = new PlayerLoopSystem[loopSystem.subSystemList.Length + 2];
			Array.Copy(loopSystem.subSystemList,0,dest,2,loopSystem.subSystemList.Length);
			dest[0] = yieldLoop;
			dest[1] = runnerLoop;
			return dest;
		}
```
